### PR TITLE
Implement BadRequestException

### PR DESCRIPTION
**Pull request by dthongvl**

### CODE REVIEW

File: javax/ws/rs/BadRequestException/BadRequestException.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "BadRequestException.hpp"
+
+using namespace Javax::Ws::Rs;
+
+BadRequestException::BadRequestException() : Exception() {
+};
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.cpp]: Please remove `;` after a `)`.

File: javax/ws/rs/BadRequestException/BadRequestException.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+BadRequestException::BadRequestException(String message) : Exception(message) {
+};
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.cpp]: Please remove `;` after a `)`.

File: javax/ws/rs/BadRequestException/BadRequestException.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+BadRequestException::BadRequestException(String message, Throwable *cause) : Exception(message, cause) {
+};
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.cpp]: Please remove `;` after a `)`.

File: javax/ws/rs/BadRequestException/BadRequestException.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+BadRequestException::BadRequestException(Throwable *cause) : Exception(cause) {
+};
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.cpp]: Please remove `;` after a `)`.

File: javax/ws/rs/BadRequestException/BadRequestException.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+using namespace Javax::Ws::Rs;
+
+BadRequestException::BadRequestException() : Exception() {
+};
+
+BadRequestException::BadRequestException(String message) : Exception(message) {
+};
+
+BadRequestException::BadRequestException(String message, Throwable *cause) : Exception(message, cause) {
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.cpp]: Please split this line.

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NATIVE_JAVAX_WS_RS_BAD_REQUEST_EXCEPTION_HPP_
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: #ifndef header guard has wrong style, please use: JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#define NATIVE_JAVAX_WS_RS_BAD_REQUEST_EXCEPTION_HPP_
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Please use: JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+#include "../../../../java/lang/Exception/Exception.hpp"
+
+using namespace Java::Lang;
+
+namespace Javax {
+    namespace Ws  {
+        namespace Rs {
+            class BadRequestException : public Exception {
+            public:
+                BadRequestException();
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Please add comments for all constructor methods.

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                BadRequestException(String message);
+                BadRequestException(String message, Throwable *cause);
+                BadRequestException(Throwable *cause);
+            };  // class BadRequestException
+        } // namespace Rs
+    }  // namespace Ws
+}  // namespace Javax
+
+
+#endif //NATIVE_JAVAX_WS_RS_BAD_REQUEST_EXCEPTION_HPP_
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Please use: JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP

Comment (ghost) [reply javax/ws/rs/BadRequestException/BadRequestException.hpp]: Should have a space between `//` and comment

Comment (ghost) [reply javax/ws/rs/BadRequestException/BadRequestException.hpp]: At least two spaces is best between code and comments.

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+#include "BadRequestException.hpp"
+
+extern "C" {
+#include "../../../../kernel/test.h"
+}
+
+using namespace Javax::Ws::Rs;
+
+TEST (BadRequestException, Constructor) {
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please replace `TEST (` by `TEST(`

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+               badRequestExceptionWithMessageAndCause.getCause()->getMessage().toString());
+
+    // Constructs a new BadRequestException with the specified cause.
+    BadRequestException badRequestExceptionWithCause = BadRequestException(&badRequestExceptionWithMessageAndCause);
+    ASSERT_STR("Illegal param", badRequestExceptionWithCause.getCause()->getMessage().toString());
+    ASSERT_STR("Illegal param",
+               badRequestExceptionWithCause.getCause()->getCause()->getMessage().toString());
+}
+
+TEST (BadRequestException, TryCatch) {
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please replace *TEST (* by *TEST(*

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    ASSERT_STR("", badRequestException.getMessage().toString());
+
+    // Constructs a new BadRequestException with the specified detail message.
+    BadRequestException badRequestExceptionWithMessage = BadRequestException(
+            "Illegal param");
+    ASSERT_STR("Illegal param",
+               badRequestExceptionWithMessage.getMessage().toString());
+
+    // Constructs a new BadRequestException with the specified detail message and cause.
+    BadRequestException badRequestExceptionWithMessageAndCause = BadRequestException("Illegal param",
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please split this line.

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                                                                                  &badRequestExceptionWithMessage);
+    ASSERT_STR("Illegal param", badRequestExceptionWithMessageAndCause.getMessage().toString());
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please split this line.

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    ASSERT_STR("Illegal param",
+               badRequestExceptionWithMessageAndCause.getCause()->getMessage().toString());
+
+    // Constructs a new BadRequestException with the specified cause.
+    BadRequestException badRequestExceptionWithCause = BadRequestException(&badRequestExceptionWithMessageAndCause);
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please split this line.

File: javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    ASSERT_STR("Illegal param", badRequestExceptionWithCause.getCause()->getMessage().toString());
+    ASSERT_STR("Illegal param",
+               badRequestExceptionWithCause.getCause()->getCause()->getMessage().toString());
+}
+
+TEST (BadRequestException, TryCatch) {
+    // Throw BadRequestException with message "throw BadRequestException", expected message "throw BadRequestException"
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestExceptionTest.cpp]: Please split this line.

**Review by ghost — CHANGES_REQUESTED** [review]
Well done bro. Just need to fix these things below to improve your source code.

**Review by ghost — APPROVED** [review]

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+#define JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+                BadRequestException(String message);
+                BadRequestException(String message, Throwable *cause);
+                BadRequestException(Throwable *cause);
+            };  // class BadRequestException
+        } // namespace Rs
+    }  // namespace Ws
+}  // namespace Javax
+
+
+#endif // JAVAX_WS_RS_BADREQUESTEXCEPTION_BADREQUESTEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/ClientErrorException/ClientErrorException.hpp [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef JAVAX_WS_RS_CLIENTERROREXCEPTION_CLIENTERROREXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/ClientErrorException/ClientErrorException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/ClientErrorException/ClientErrorException.hpp [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+#define JAVAX_WS_RS_CLIENTERROREXCEPTION_CLIENTERROREXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/ClientErrorException/ClientErrorException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/ClientErrorException/ClientErrorException.hpp [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+                ClientErrorException(String message);
+                ClientErrorException(String message, Throwable *cause);
+                ClientErrorException(Throwable *cause);
+            };  // class ClientErrorException
+        } // namespace Rs
+    }  // namespace Ws
+}  // namespace Javax
+
+
+#endif // JAVAX_WS_RS_CLIENTERROREXCEPTION_CLIENTERROREXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/ClientErrorException/ClientErrorException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/WebApplicationException/WebApplicationException.hpp [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef JAVAX_WS_RS_WEBAPPLICATIONEXCEPTION_WEBAPPLICATIONEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/WebApplicationException/WebApplicationException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/WebApplicationException/WebApplicationException.hpp [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+#define JAVAX_WS_RS_WEBAPPLICATIONEXCEPTION_WEBAPPLICATIONEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/WebApplicationException/WebApplicationException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/WebApplicationException/WebApplicationException.hpp [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+                WebApplicationException(String message);
+                WebApplicationException(String message, Throwable *cause);
+                WebApplicationException(Throwable *cause);
+            };  // class WebApplicationException
+        } // namespace Rs
+    }  // namespace Ws
+}  // namespace Javax
+
+
+#endif // JAVAX_WS_RS_WEBAPPLICATIONEXCEPTION_WEBAPPLICATIONEXCEPTION_HPP
```
Comment (ghost) [inline javax/ws/rs/WebApplicationException/WebApplicationException.hpp]: Missing `_` in the end of line.

File: javax/ws/rs/BadRequestException/BadRequestException.hpp [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+namespace Javax {
+    namespace Ws  {
+        namespace Rs {
+            class BadRequestException : public ClientErrorException {
+            public:
+                BadRequestException();
+                BadRequestException(String message);
+                BadRequestException(String message, Throwable *cause);
+                BadRequestException(Throwable *cause);
+            };  // class BadRequestException
```
Comment (ghost) [inline javax/ws/rs/BadRequestException/BadRequestException.hpp:43]: How about implementing `getCause()`, `getMessage()` method?

**Review by ghost — CHANGES_REQUESTED** [review]

File: javax/ws/rs/BadRequestException/BadRequestException.hpp (hunk 25 above)
Comment (dthongvl) [reply javax/ws/rs/BadRequestException/BadRequestException.hpp:43]: Inheriting from the superclass Throwable

**Review by ghost — APPROVED** [review]

**Pull request merged**